lib/run/frame.ts

```diff
@@ -17,83 +18,182 @@ export interface FrameOptions<T> {
 }
 
 export function createFrame<T>(options: FrameOptions<T>): Frame<T> {
-  let { operation, parent } = options;
-  let children = new Set<Frame>();
-  let block = createBlock(operation);
-  let context = Object.create(parent?.context ?? {});
-  let results = createEventStream<void, Result<void>>();
+  return evaluate<Frame<T>>(function* () {
+    let { operation, parent } = options;
+    let children = new Set<Frame>();
+    let context = Object.create(parent?.context ?? {});
+    let thunks: IteratorResult<Thunk, Result<T>>[] = [{
+      done: false,
+      value: $next(void 0),
+    }];
+
+    let crash: Error | undefined = void 0;
+
+    let interrupt = () => {};
+
+    let abort = (reason?: Error) => {
+      if (!frame.aborted) {
+        frame.aborted = true;
+        crash = reason;
+        thunks.unshift({ done: false, value: $abort() });
+        interrupt();
+      }
+    };
+
+    let [setResults, results] = yield* createValue<FrameResult<T>>();
+
+    let frame = yield* shiftSync<Frame<T>>((k) => {
+      let self: Frame<T> = create<Frame<T>>("Frame", { id: ids++, context }, {
+        createChild<X>(operation: () => Operation<X>) {
+          let child = createFrame<X>({ operation, parent: frame });
+          children.add(child);
+          evaluate(function* () {
+            yield* child;
+            children.delete(child);
+          });
```

Comment:
If `child` is asynchronous then that means this evaluate function will complete later and then the child will be deleted at some time in the future, right?

Reply:
Yes, it means that however the child resolves, delete that child from the children immediately. So, for example, if a spawned task finishes, it will be removed after it completes.